bee-message/src/address/mod.rs

```diff
@@ -0,0 +1,114 @@
+// Copyright 2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+mod bls;
+mod ed25519;
+
+pub use bls::{BlsAddress, BLS_ADDRESS_LENGTH};
+pub use ed25519::{Ed25519Address, ED25519_ADDRESS_LENGTH};
+
+use crate::{error::ValidationError, signature::Signature};
+
+use bee_packable::Packable;
+
+use bech32::{self, FromBase32, ToBase32, Variant};
+
+use alloc::{str::FromStr, string::String, vec::Vec};
+use core::convert::TryFrom;
+
+/// A generic address supporting different address kinds.
+#[derive(Clone, Debug, Eq, PartialEq, Ord, PartialOrd, Hash, Packable)]
+#[cfg_attr(
+    feature = "serde1",
+    derive(serde::Serialize, serde::Deserialize),
+    serde(tag = "type", content = "data")
+)]
+#[packable(tag_type = u8)]
+pub enum Address {
+    /// An Ed25519 address.
+    #[packable(tag = 0)]
+    Ed25519(Ed25519Address),
+    /// A BLS address.
+    #[packable(tag = 1)]
+    Bls(BlsAddress),
+}
+
+impl Address {
+    /// Returns the address kind of an `Address`.
+    pub fn kind(&self) -> u8 {
+        match self {
+            Self::Ed25519(_) => Ed25519Address::KIND,
+            Self::Bls(_) => BlsAddress::KIND,
+        }
+    }
+
+    /// Tries to create an `Address` from a Bech32 encoded string.
+    pub fn try_from_bech32(addr: &str) -> Result<Self, ValidationError> {
```

Comment:
that we also decode an `Address` from `bech32` strengthens my opinion on the `FromStr` impls for the concrete address types. So for them we can only use `base16`, but on the enum only `bech32`?